### PR TITLE
WIP: Auto Flexvolume plugin discovery, with prober dependency injection through VolumePluginMgr initializer.

### PR DESCRIPTION
**Pull request by verult**
See the design proposal here: kubernetes/community#833, and previous prototype PR #49600

This is a different prototype for injecting the Prober dependency. I aimed to have better package separation with this implementation.

do-not-merge

/release-note-none

/sig-storage

/assign @chakri-nelluri @saad-ali @gnufied

### CODE REVIEW

File: cmd/kube-controller-manager/app/plugins.go [hunk 1]
```diff
@@ -65,13 +65,13 @@ import (
 // plugins.
 // The list of plugins is manually compiled. This code and the plugin
 // initialization code for kubelet really, really need a through refactor.
-func ProbeAttachableVolumePlugins(config componentconfig.VolumeConfiguration) []volume.VolumePlugin {
+func ProbeAttachableVolumePlugins() []volume.VolumePlugin {
```
Comment (php-coder) [inline cmd/kube-controller-manager/app/plugins.go]: Could you also update the comment?

File: cmd/kubelet/app/plugins.go [hunk 2]
```diff
@@ -63,7 +63,7 @@ import (
 // ProbeVolumePlugins collects all volume plugins into an easy to use list.
 // PluginDir specifies the directory to search for additional third party
 // volume plugins.
-func ProbeVolumePlugins(pluginDir string) []volume.VolumePlugin {
+func ProbeVolumePlugins() []volume.VolumePlugin {
```
Comment (php-coder) [inline cmd/kubelet/app/plugins.go]: The comment is outdated now. Please, update it too.

File: cmd/kube-controller-manager/app/plugins.go (hunk 1 above)
Comment (verult) [inline cmd/kube-controller-manager/app/plugins.go]: ACK. Will update on the actual complete PR.

File: pkg/controller/volume/persistentvolume/pv_controller_base.go [hunk 3]
```diff
@@ -89,7 +89,7 @@ func NewController(p ControllerParameters) (*PersistentVolumeController, error)
 		volumeQueue:                   workqueue.NewNamed("volumes"),
 	}
 
-	if err := controller.volumePluginMgr.InitPlugins(p.VolumePlugins, controller); err != nil {
+	if err := controller.volumePluginMgr.InitPlugins(p.VolumePlugins, nil, controller); err != nil {
```
Comment (saad-ali) [inline pkg/controller/volume/persistentvolume/pv_controller_base.go]: Why is prober nil? 

Also wondering if you can do `volumePluginMgr.InitPlugins(...)` at layer above and have both PV/PVC controller and attach/detach controller share a volumePluginMgr. <- Don't worry about this for this PR, can do a separate refactor PR 
`

**Review by saad-ali — COMMENTED** [review]
Don't have a strong preference between this and https://github.com/kubernetes/kubernetes/pull/49600

File: pkg/controller/volume/persistentvolume/pv_controller_base.go (hunk 3 above)
Comment (verult) [inline pkg/controller/volume/persistentvolume/pv_controller_base.go]: PV/PVC controller doesn't have Flexvolume plugin registered.

**Comment by verult** [comment]
Implemented as part of #50031

**Pull request closed without merging**